contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+func (t *tracerImpl) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	var carrier tracer.TextMapCarrier = m
+	ctx, err := tracer.Extract(carrier)
+	if err != nil {
+		if errors.Is(err, tracer.ErrSpanContextNotFound) {
+			// If there is no span, return nothing, but don't error out. This is
+			// a legitimate place where a span does not exist in the headers
+			return nil, nil
+		}
+		return nil, err
+	}
+	return &tracerSpanCtx{ctx}, nil
+}
+
+func (t *tracerImpl) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	carrier := tracer.TextMapCarrier{}
+	if err := tracer.Inject(span.(*tracerSpan).Context(), carrier); err != nil {
+		return nil, err
+	}
+	return carrier, nil
+}
+
+func (t *tracerImpl) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span, ok := tracer.SpanFromContext(ctx)
+	if !ok {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (t *tracerImpl) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return tracer.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func genSpanID(idempotencyKey string) uint64 {
+	h := fnv.New64()
+	// Write() always writes all bytes and never fails; the count and error result are for implementing io.Writer.
+	_, _ = h.Write([]byte(idempotencyKey))
+	return h.Sum64()
+}
+
+func (t *tracerImpl) StartSpan(options *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	startOpts := []tracer.StartSpanOption{
+		tracer.ResourceName(options.Name),
+		tracer.StartTime(options.Time),
+	}
+	// Set a deterministic span ID for workflows which are long-running and cross process boundaries
+	if options.Operation == "RunWorkflow" {
+		startOpts = append(startOpts, tracer.WithSpanID(genSpanID(options.IdempotencyKey)))
+	}
+
+	// Add parent span to start options
+	var parent ddtrace.SpanContext
+	switch opParent := options.Parent.(type) {
+	case nil:
+	case *tracerSpan:
+		parent = opParent.Context()
+	case *tracerSpanCtx:
+		parent = opParent
+	default:
+		return nil, fmt.Errorf("unrecognized parent type %T", opParent)
+	}
+	if parent != nil {
+		// Convert the parent context into a full trace.
+		// This is required, otherwise the implementation of StartSpan()
+		// will NOT respect ChildOf().
+		// The runtime type of parentTrace will be tracer.spanContext (note the
+		// starting lowercase on "spanContext", that's an internal struct)
+		parentTrace, err := tracer.Extract(newSpanContextReader(parent))
+		if err != nil {
+			return nil, err
+		}
+		startOpts = append(startOpts, tracer.ChildOf(parentTrace))
+	}
+
+	// Add tags to start options
+	for k, v := range options.Tags {
+		// Display Temporal tags in a nested group in Datadog APM
+		tagKey := "temporal." + strings.TrimPrefix(k, "temporal")
+		startOpts = append(startOpts, tracer.Tag(tagKey, v))
+	}
+
+	// Start and return span
+	s := tracer.StartSpan(options.Operation+":"+options.Name, startOpts...)
```

Comment:
We don't include `options.Name` in the span's operation type in our internal implementation. Was there a reason to do this here? It's already set as the span's resource name at the top of the function.

Internally we also manually add a `temporal.` prefix in order to contextualize spans (names like `ExecuteWorkflow` are very generic when paired next to tens or hundreds of other span types).

```suggestion
	s := tracer.StartSpan("temporal."+options.Operation, startOpts...)
```

Reply:
The only reason is that the current testing suite expects the operation names to follow a given pattern. I do agree that we should detach the naming pattern or at least make it configurable somehow, and the same goes for adding a prefix (that should default to temporal).

Reply:
I think we'll need to update the test suite then, because separately indexing span types like `temporal.RunWorkflow` and [dynamic resources](https://docs.datadoghq.com/tracing/glossary/#resources) like `MyWorkflowName` is key to how we expect customers to set up generic monitoring for things like per-workflow-type error rate or execution latency.

Let's see if we can customize the tests to enforce the current behavior for the other tracing implementations, but allow a custom prefix and span names for Datadog. We'll need another review from @cretz on that of course.

Reply:
:+1: I would support altering the shared test in a way that lets this work while still letting the test control what spans its asserting presence of and letting the other tests still pass.